File: engine/platform/ios/app/Ads/LevelPlayAdsBackend.swift
// =============================================================================
//  Aurea iOS — Unity LevelPlay (SDK 9.6.0, API `LPM*`). O provedor ATIVO no
//  iOS, como no Android. Só este arquivo importa o SDK do LevelPlay.
//
//   - consentimento pelo UMP (TCF) antes de tudo; depois `LevelPlay.initWith`,
//     e os anúncios só são criados/carregados depois do init com sucesso;
//   - Rewarded: a recompensa vem SÓ de `didRewardAd`; `didCloseAd` é só fechar;
//   - App Open: não existe no LevelPlay — este backend não carrega App Open;
//   - a rede Unity Ads (Game ID/placements) é configurada no painel do
//     LevelPlay; o app só conhece o App Key e as unidades.
// =============================================================================
import UIKit
import IronSource

final class LevelPlayAdsBackend: NSObject, AdsBackend {
    private let appKey: String
    private let debug: Bool
    /// DEBUG: abre a Integration Test Suite oficial assim que o init terminar.
    private let openTestSuite: Bool

    private var interstitial: LPMInterstitialAd?
    private var rewardedUnit = ""
    private var rewardedAd: LPMRewardedAd?
    private var interLoad: (ok: () -> Void, fail: (String) -> Void)?
    private var rewardLoad: (ok: () -> Void, fail: (String) -> Void)?
    private var interShow: ShowCallbacks?
    private var rewardShow: ShowCallbacks?

    private struct ShowCallbacks {
        let opened: () -> Void, dismissed: () -> Void, failed: (String) -> Void, rewarded: () -> Void
    }

    init(appKey: String, debug: Bool, openTestSuite: Bool = false) {
        self.appKey = appKey; self.debug = debug; self.openTestSuite = openTestSuite
    }

    private func onMain(_ block: @escaping () -> Void) {
        if Thread.isMainThread { block() } else { DispatchQueue.main.async(execute: block) }
    }

    func initialize(from host: AnyObject, done: @escaping (Bool) -> Void) {
        guard let vc = host as? UIViewController else { done(false); return }
        var answered = false
        let finish: (Bool) -> Void = { [weak self] ok in self?.onMain { if !answered { answered = true; done(ok) } } }
        UmpConsent.request(from: vc) { [weak self] canRequest in
            guard let self else { return }
            guard canRequest else { finish(false); return }
            if self.debug {
                // Ferramenta oficial de teste do LevelPlay (Integration Test Suite).
                LevelPlay.setMetaDataWithKey("is_test_suite", value: "enable")
            }
            let request = LPMInitRequestBuilder(appKey: self.appKey).build()
            LevelPlay.initWith(request) { [weak self] _, error in
                guard let self else { return }
                if error != nil { finish(false); return }
                if self.debug && self.openTestSuite {
                    self.onMain { LevelPlay.launchTestSuite(vc) }
                }
                finish(true)
            }
        }
    }

    func load(_ kind: AdKind, unitId: String, loaded: @escaping () -> Void, failed: @escaping (String) -> Void) {
        switch kind {
        case .appOpen:
            onMain { failed("App Open não existe no LevelPlay") }
        case .exportInterstitial:
            if interstitial == nil {
                let a = LPMInterstitialAd(adUnitId: unitId)
                a.setDelegate(self)
                interstitial = a
            }
            guard let ad = interstitial else { return }
            interLoad = (loaded, failed)
            ad.loadAd()
        case .aiRewarded:
            if rewardedAd == nil {
                let a = LPMRewardedAd(adUnitId: unitId)
                rewardedUnit = unitId
                a.setDelegate(self)
                rewardedAd = a
            }
            guard let ad = rewardedAd else { return }
            rewardLoad = (loaded, failed)
            ad.loadAd()
        }
    }

    func show(_ kind: AdKind, from host: AnyObject, opened: @escaping () -> Void,
              dismissed: @escaping () -> Void, failed: @escaping (String) -> Void,
              rewarded: @escaping () -> Void) -> Bool {
        guard var vc = host as? UIViewController else { return false }
        // Por cima de qualquer folha aberta (ex.: a da exportação).
        while let top = vc.presentedViewController, !top.isBeingDismissed { vc = top }
        guard vc.viewIfLoaded?.window != nil else { return false }
        let cb = ShowCallbacks(opened: opened, dismissed: dismissed, failed: failed, rewarded: rewarded)
        switch kind {
        case .exportInterstitial:
            guard let ad = interstitial, ad.isAdReady() else { return false }
            interShow = cb
            ad.showAd(viewController: vc, placementName: nil)
            return true
        case .aiRewarded:
            guard let ad = rewardedAd, ad.isAdReady() else { return false }
            rewardShow = cb
            ad.showAd(viewController: vc, placementName: nil)
            return true
        case .appOpen:
            return false
        }
    }

    /// O objeto do LevelPlay é reaproveitado (carregar de novo depois de mostrar); só os callbacks saem.
    func release(_ kind: AdKind) {
        switch kind {
        case .exportInterstitial: interLoad = nil
        case .aiRewarded: rewardLoad = nil
        case .appOpen: break
        }
    }

    fileprivate static func describe(_ error: Error) -> String {
        let e = error as NSError
        return "levelplay \(e.code): \(e.localizedDescription)"
    }
}

extension LevelPlayAdsBackend: LPMInterstitialAdDelegate {
    func didLoadAd(with adInfo: LPMAdInfo) {
        #if DEBUG
        print("[AUREA ADS] LevelPlay carregou \(adInfo.adUnitId) pela rede \(adInfo.adNetwork)")
        #endif
        onMain { [weak self] in
            // O mesmo delegate atende aos dois formatos: decide pelo ID da unidade.
            guard let self else { return }
            if adInfo.adUnitId == self.rewardedUnit {
                self.rewardLoad?.ok(); self.rewardLoad = nil
            } else {
                self.interLoad?.ok(); self.interLoad = nil
            }
        }
    }
    func didFailToLoadAd(withAdUnitId adUnitId: String, error: Error) {
        let msg = Self.describe(error)
        onMain { [weak self] in
            guard let self else { return }
            if adUnitId == self.rewardedUnit {
                self.rewardLoad?.fail(msg); self.rewardLoad = nil
            } else {
                self.interLoad?.fail(msg); self.interLoad = nil
            }
        }
    }
    func didDisplayAd(with adInfo: LPMAdInfo) {
        onMain { [weak self] in
            guard let self else { return }
            (adInfo.adUnitId == self.rewardedUnit ? self.rewardShow : self.interShow)?.opened()
        }
    }
    func didFailToDisplayAd(with adInfo: LPMAdInfo, error: Error) {
        let msg = Self.describe(error)
        onMain { [weak self] in
            guard let self else { return }
            if adInfo.adUnitId == self.rewardedUnit {
                self.rewardShow?.failed(msg); self.rewardShow = nil
            } else {
                self.interShow?.failed(msg); self.interShow = nil
            }
        }
    }
    /// Fechar NÃO é recompensa.
    func didCloseAd(with adInfo: LPMAdInfo) {
        onMain { [weak self] in
            guard let self else { return }
            if adInfo.adUnitId == self.rewardedUnit {
                self.rewardShow?.dismissed(); self.rewardShow = nil
            } else {
                self.interShow?.dismissed(); self.interShow = nil
            }
        }
    }
}

extension LevelPlayAdsBackend: LPMRewardedAdDelegate {
    /// A ÚNICA recompensa.
    func didRewardAd(with adInfo: LPMAdInfo, reward: LPMReward) {
        onMain { [weak self] in self?.rewardShow?.rewarded() }
    }
}

File: android/app/src/main/java/com/aurea/aurea/ads/LevelPlayAdsBackend.kt
package com.aurea.aurea.ads

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.unity3d.mediation.LevelPlay
import com.unity3d.mediation.LevelPlayAdError
import com.unity3d.mediation.LevelPlayAdInfo
import com.unity3d.mediation.LevelPlayConfiguration
import com.unity3d.mediation.LevelPlayInitError
import com.unity3d.mediation.LevelPlayInitListener
import com.unity3d.mediation.LevelPlayInitRequest
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAd
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAdListener
import com.unity3d.mediation.rewarded.LevelPlayReward
import com.unity3d.mediation.rewarded.LevelPlayRewardedAd
import com.unity3d.mediation.rewarded.LevelPlayRewardedAdListener
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Unity LevelPlay (API atual, `com.unity3d.mediation`) — o provedor ativo no
 * Android. Só este arquivo importa o SDK do LevelPlay.
 *
 *  - consentimento pelo UMP (TCF) antes de tudo; depois `LevelPlay.init`, e os
 *    anúncios só são criados/carregados depois de `onInitSuccess`;
 *  - Rewarded: a recompensa vem SÓ de `onAdRewarded`; `onAdClosed` é só fechar;
 *  - App Open: não existe no LevelPlay — este backend não carrega App Open;
 *  - os callbacks do SDK voltam para a thread principal (o manager é single-thread).
 */
class LevelPlayAdsBackend(
    context: Context,
    private val appKey: String,
    private val debug: Boolean,
    /** DEBUG: abre a Test Suite oficial do LevelPlay assim que o init terminar. */
    private val abrirTestSuite: Boolean = false,
) : AdsBackend {

    private val app = context.applicationContext
    private val main = Handler(Looper.getMainLooper())
    private fun naMain(bloco: () -> Unit) = main.post { runCatching(bloco) }

    private var interstitial: LevelPlayInterstitialAd? = null
    private var rewarded: LevelPlayRewardedAd? = null

    // Callbacks do carregamento/exibição em curso (um de cada por vez).
    private var interLoad: Pair<() -> Unit, (String) -> Unit>? = null
    private var rewardLoad: Pair<() -> Unit, (String) -> Unit>? = null
    private var interShow: ShowCallbacks? = null
    private var rewardShow: ShowCallbacks? = null

    private class ShowCallbacks(
        val aoAbrir: () -> Unit, val aoFechar: () -> Unit, val aoFalhar: (String) -> Unit, val aoRecompensar: () -> Unit,
    )

    override fun initialize(host: Any, aoTerminar: (podePedir: Boolean) -> Unit) {
        val activity = host as? Activity ?: run { aoTerminar(false); return }
        val uma = AtomicBoolean(false)
        val terminar = { ok: Boolean -> if (uma.compareAndSet(false, true)) naMain { aoTerminar(ok) } }
        UmpConsent.pedir(activity) { podePedir ->
            if (!podePedir) { terminar(false); return@pedir }
            try {
                if (debug) {
                    LevelPlay.setAdaptersDebug(true)
                    // Ferramenta oficial de teste do LevelPlay (Integration Test Suite).
                    LevelPlay.setMetaData("is_test_suite", "enable")
                }
                val pedido = LevelPlayInitRequest.Builder(appKey).build()
                LevelPlay.init(app, pedido, object : LevelPlayInitListener {
                    override fun onInitSuccess(configuration: LevelPlayConfiguration) {
                        if (debug && abrirTestSuite) naMain { runCatching { LevelPlay.launchTestSuite(app) } }
                        terminar(true)
                    }
                    override fun onInitFailed(error: LevelPlayInitError) {
                        terminar(false)
                    }
                })
            } catch (t: Throwable) {
                terminar(false)
            }
        }
    }

    override fun loadAppOpen(unitId: String, aoCarregar: () -> Unit, aoFalhar: (String) -> Unit) {
        naMain { aoFalhar("App Open não existe no LevelPlay") }
    }

    override fun loadInterstitial(unitId: String, aoCarregar: () -> Unit, aoFalhar: (String) -> Unit) {
        val ad = interstitial ?: LevelPlayInterstitialAd(unitId).also {
            it.setListener(interListener)
            interstitial = it
        }
        interLoad = aoCarregar to aoFalhar
        ad.loadAd()
    }

    override fun loadRewarded(unitId: String, aoCarregar: () -> Unit, aoFalhar: (String) -> Unit) {
        val ad = rewarded ?: LevelPlayRewardedAd(unitId).also {
            it.setListener(rewardListener)
            rewarded = it
        }
        rewardLoad = aoCarregar to aoFalhar
        ad.loadAd()
    }

    override fun show(kind: AdKind, host: Any, aoAbrir: () -> Unit, aoFechar: () -> Unit, aoFalhar: (String) -> Unit,
                      aoRecompensar: () -> Unit): Boolean {
        val activity = host as? Activity ?: return false
        if (activity.isFinishing || activity.isDestroyed) return false
        val cb = ShowCallbacks(aoAbrir, aoFechar, aoFalhar, aoRecompensar)
        return when (kind) {
            AdKind.ExportInterstitial -> {
                val ad = interstitial?.takeIf { it.isAdReady() } ?: return false
                interShow = cb
                ad.showAd(activity)
                true
            }
            AdKind.AiRewarded -> {
                val ad = rewarded?.takeIf { it.isAdReady() } ?: return false
                rewardShow = cb
                ad.showAd(activity)
                true
            }
            AdKind.AppOpen -> false
        }
    }

    /** Dynamic User ID: chega no callback S2S como [USER_ID] (precisa vir ANTES do show). */
    override fun setRewardUserId(id: String) {
        runCatching { LevelPlay.setDynamicUserId(id) }
    }

    /** O objeto do LevelPlay é reaproveitado (carregar de novo depois de mostrar); só os callbacks saem. */
    override fun release(kind: AdKind) {
        when (kind) {
            AdKind.ExportInterstitial -> interLoad = null
            AdKind.AiRewarded -> rewardLoad = null
            AdKind.AppOpen -> Unit
        }
    }

    private fun erro(e: LevelPlayAdError) = "levelplay ${e.getErrorCode()}: ${e.getErrorMessage()}"

    private val interListener = object : LevelPlayInterstitialAdListener {
        override fun onAdLoaded(adInfo: LevelPlayAdInfo) {
            naMain { interLoad?.first?.invoke(); interLoad = null }
        }
        override fun onAdLoadFailed(error: LevelPlayAdError) {
            naMain { interLoad?.second?.invoke(erro(error)); interLoad = null }
        }
        override fun onAdDisplayed(adInfo: LevelPlayAdInfo) {
            naMain { interShow?.aoAbrir?.invoke() }
        }
        override fun onAdDisplayFailed(error: LevelPlayAdError, adInfo: LevelPlayAdInfo) {
            naMain { interShow?.aoFalhar?.invoke(erro(error)); interShow = null }
        }
        override fun onAdClosed(adInfo: LevelPlayAdInfo) {
            naMain { interShow?.aoFechar?.invoke(); interShow = null }
        }
    }

    private val rewardListener = object : LevelPlayRewardedAdListener {
        override fun onAdLoaded(adInfo: LevelPlayAdInfo) {
            naMain { rewardLoad?.first?.invoke(); rewardLoad = null }
        }
        override fun onAdLoadFailed(error: LevelPlayAdError) {
            naMain { rewardLoad?.second?.invoke(erro(error)); rewardLoad = null }
        }
        override fun onAdDisplayed(adInfo: LevelPlayAdInfo) {
            naMain { rewardShow?.aoAbrir?.invoke() }
        }
        override fun onAdDisplayFailed(error: LevelPlayAdError, adInfo: LevelPlayAdInfo) {
            naMain { rewardShow?.aoFalhar?.invoke(erro(error)); rewardShow = null }
        }
        /** A ÚNICA recompensa. */
        override fun onAdRewarded(reward: LevelPlayReward, adInfo: LevelPlayAdInfo) {
            naMain { rewardShow?.aoRecompensar?.invoke() }
        }
        /** Fechar NÃO é recompensa. */
        override fun onAdClosed(adInfo: LevelPlayAdInfo) {
            naMain { rewardShow?.aoFechar?.invoke(); rewardShow = null }
        }
    }
}
